base/src/commonMain/kotlin/state/State.kt

```diff
@@ -20,57 +20,91 @@ Copyright 2019 Splendo Consulting B.V. The Netherlands
 import com.splendo.kaluga.flow.BaseFlowable
 import com.splendo.kaluga.base.runBlocking
 import kotlinx.coroutines.*
+import kotlin.coroutines.CoroutineContext
 
-open class State<T:State<T>>(open val repo:StateRepo<T>){
+open class State<T:State<T>>(open val repoAccessor:StateRepoAccesor<T>){
+    open suspend fun initialState() {}
     open suspend fun beforeCreatingNewState() {}
-    open suspend fun afterCreatingNewState() {}
+    open suspend fun afterCreatingNewState(newState: T) {}
     open suspend fun afterNewStateIsSet() {}
+    open suspend fun beforeOldStateIsRemoved() {}
+    open suspend fun afterOldStateIsRemoved(oldState: T) {}
+    open suspend fun finalState() {}
+
+    protected suspend fun changeState(toState: T) {
+        repoAccessor.changeState {
+            if (it === this)
+                toState
+            else
+                it
+        }
+    }
+
+}
+
+class StateRepoAccesor<T:State<T>>(private val s:StateRepo<T> ) : CoroutineScope by s {
+
+    fun currentState(): T {
```

Comment:
Can be simple: `fun currentState() = s.state()` as it is in Coding Conventions of Kotlin